### PR TITLE
Add ffmpeg support to workflow migration script

### PR DESCRIPTION
**Pull request by dichaudhary**
## Description

PR contains changes to include com.day.cq.dam.video.FFMpegTranscodeProcess as NUI supports it

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by ireasor — CHANGES_REQUESTED** [review]
Unfortunately, our options for video encoding on Cloud Service is much more limited than what was available when running FFMpeg from a workflow.  If you take a look at the processing profile UI, you'll notice that we only support mp4 as an output format and only support h264 as a codec.

In your changes, it seems like we are taking whatever the user has and are mapping it to a processing profile, but this won't work for any cases other than h264 mp4 files.  We should check and ensure that the format will be supported before creating the processing profile and we should warn the user if their configurations fall outside of these parameters.

It would be a good idea to take any processing profile that is output and test it out to ensure that it works properly.  For example, the ogg settings that your test generates will not work properly on a real environment.

**Comment by dichaudhary** [comment]
> Unfortunately, our options for video encoding on Cloud Service is much more limited than what was available when running FFMpeg from a workflow. If you take a look at the processing profile UI, you'll notice that we only support mp4 as an output format and only support h264 as a codec.
> 
> In your changes, it seems like we are taking whatever the user has and are mapping it to a processing profile, but this won't work for any cases other than h264 mp4 files. We should check and ensure that the format will be supported before creating the processing profile and we should warn the user if their configurations fall outside of these parameters.
> 
> It would be a good idea to take any processing profile that is output and test it out to ensure that it works properly. For example, the ogg settings that your test generates will not work properly on a real environment.

Hi @ireasor , thanks for the feedback, Actually in the start I was aware about the limitations but as I progressed the changes became huge that I forgot to put the checks , I have updated the checks and test cases both. thanks again

**Comment by ireasor** [comment]
Thanks, @dichaudhary.  Unfortunately, I don't think it is as simple as using Logger.WARN.  We output a report at the end of the execution with the steps that we have taken and any information that the customer needs to know.  It is important that we include this information there.  Other parts of the code accomplish this via the ChangeTrackingService and the MigrationReportWriter.

File: src/main/java/com/adobe/skyline/migration/transformer/processingprofile/mappers/FFMpegTranscodeProcessMapper.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        } catch (Exception e) {
+            Logger.ERROR("Unable to parse details for FFMpegTranscodeProcess, make sure that video config file at " + videoProfilePath + " s present and in correct format " + e.getMessage());
+            return new ArrayList<>();
+        }
+        return renditions;
+    }
+
+    private RenditionConfig getRendition(String config, String videoProfilePath) throws MigrationRuntimeException{
+        String[] tokens = config.split(":");
+        String profile = tokens[1].trim();
```
Comment (vaibhav2601) [inline src/main/java/com/adobe/skyline/migration/transformer/processingprofile/mappers/FFMpegTranscodeProcessMapper.java:103]: Incase tokens array is null ? Will it be okay to check the size ?

**Review by ireasor — COMMENTED** [review]
The behavior is mostly correct now, but I worry that simply adding the processing profile to the report with a "Failed" flag is not going to mean anything to the user.

Additionally, adding the ProcessingProfile object to the ProfileMapper interface and then passing in null for all of the existing implementations of it feels like a violation of the Interface Segregation Principle of Object Oriented Design.

I think it would be better to add a throws clause to the interface, maybe creating a "ProfileMappingException" and use that to throw an exception when you encounter a video profile that cannot be mapped.  Then, in the ProcessingProfileCreator, you can log the video profile to the ChangeTrackingService.  Finally, when outputting the report, you can add a message to the log that the video profile at <PATH> could not be mapped to a processing profile.  We should ideally point them to documentation so they can learn why we cannot support this video profile.  If it's possible to add that to the exception, and thus what we output in the report, that would be even better.  Maybe something like "ogg format not supported".

**Comment by dichaudhary** [comment]
@ireasor, The only thought process behind this approach was to make sure that if there are two or more profiles configured for video like in our test case (profilewithmp4, ogg) and out of which one with mp4 is supported , in such a case I wanted to make sure that processing profile for the supported one is created and the ones which are not supported will be tracked as failed profiles in FFMpegTranscodeProcessMapper itself and that's why the change tracker was passed to the mapper, If we are good to skip creating profiles for the ones which are supported for the reason even if there is one unsupported profile , I am ok to make that change , let me know your thoughts!

**Comment by ireasor** [comment]
@dichaudhary I could see that.  That being said, I still don't think that we need to pass in the processing profile here.  We are not failing to create the processing profile but rather are unable to create a rendition config for a profile from the workflow step.  If you were to instead add `trackMappingError()` to `ChangeTrackingService`, then you wouldn't need to pass the processing profile object around.  You could even log a string here that contains the failed workflow step location and the reason why it cannot be mapped.

**Comment by dichaudhary** [comment]
> We are not failing to create the processing profile but rather are unable to create a rendition config for a profile from the workflow step. If you were to instead add `trackMappingError()` to `ChangeTrackingService`, then you wouldn't need to pass the processing profile object around. 

@ireasor, Agreed, with this part! I have made the changes to the report generation and interfaces accordingly. This is how it will look in the report if there are failures for any video profile , ones which are supported will succeed !

image

**Review by ireasor — APPROVED** [review]
Thanks @dichaudhary, this looks good.  One last thing to consider is that you may want to add some logic to the MigrationOrchestratorIntegrationTest to include this use case.

**Comment by dichaudhary** [comment]
> Thanks @dichaudhary, this looks good. One last thing to consider is that you may want to add some logic to the MigrationOrchestratorIntegrationTest to include this use case.

Thanks @ireasor  , I have created a separate task to track this under [CQ-4327668](https://jira.corp.adobe.com/browse/CQ-4327668) , I will be merging this now !

**Pull request merged**